### ISSUE TITLE
Missing mechanism to pass/receive struct values into/from FFI C functions

### ISSUE THREAD

**Issue by stefandd** [ponylang/rfcs#183]
The commit https://github.com/ponylang/ponyc/commit/221169bc60414eebae2ce733a794e2fd3d5fda76 removed, on the basis of some C ABI-dependent unexpected behaviors/bugs, the only mechanism to pass value structs into FFI functions or to receive them. In the commit, a proposal was made to use libclang to access the correct call signatures and the removal of tuples for FFI functions was intended to be an interim measure until such a change was made: `"... This change is a temporary measure to avoid spurious FFI bugs until we have the interoperability detailed above. LLVM intrinsics are still allowed to return and be passed tuples since their signatures are defined by LLVM."`

A little more than 3 years later no solution is in sight, now somewhat putting into question the decision to entirely remove the ability to wrap C functions that expect or return struct values. This severely limits the ability to wrap large C libraries with fairly simple value structs such as sfVector2f below for which the tuple solution might have worked fine! There are hacks one can use to make this work for this specific struct or structs of good-natured size.

Given the usefulness of a mechanism to pass or receive value structs in FFI scenarios and given the lack of progress that would have justified the removal of the tuple mechanism from the language althogether, I propose to **either renable the passing and receiving of tuples (revert the above commit) or to provide a mechanism by which chunks of raw byte data can be generated, cast into Pony primitives, and passed into or received from FFI C functions. Unfortunately, even the hack workarounds shown below only work if by chance the struct size matches the size of a Pony primitive like U32, U64 etc. The struct sfVector3f with three floats cannot be worked around this way since there is no U96 or I96 in Pony.**

----------------------------- Detailed example

Take, for example, the following C library wrapped in Pony as "debugstruct.lib"
```c
typedef struct
{
    float x;
    float y;
} sfVector2f;

void PrintStruct(sfVector2f s) {
    printf("%f, %f\n", s.x, s.y);
}

unsigned long long sfVector2f_to_U64(sfVector2f* s) {
    return *((unsigned long long*)(s));
}
```
This currently can't be made to print the sfVector2f struct with Pony code such as:
```pony
use "lib:debugstruct"
use @PrintStruct[None](vec : SFVector2f) // does not work because it passes by reference
//use @PrintStruct[None](x : F32, y : F32)  // this does not work as work-around

struct SFVector2f
    let posx : F32
    let posy : F32

    new create(x : F32, y : F32) =>
        posx = x
        posy = y

actor Main
    let _env : Env

    new create(env : Env) =>
        _env = env
        let vec : SFVector2f = SFVector2f(1.0, 2.0)
        @PrintStruct[None](vec) // compiles but does not work
        // @PrintStruct[None](vec.x, vec.y) // using the second definition above it also compiles but does not work
```
*For this specific 64-bit struct*, this can be made work by a hack that involves both, another C function that casts the two float-struct into a unsigned long long in C, and *misinforming* Pony about the signature of the C function.
```pony
use "lib:debugstruct"
use @PrintStruct[None](arg : U64)
use @sfVector2f_to_U64[U64](arg : SFVector2f)

struct SFVector2f
    let posx : F32
    let posy : F32

    new create(x : F32, y : F32) =>
        posx = x
        posy = y

actor Main
    let _env : Env
    let vec : SFVector2f = SFVector2f(1.0, 2.0)
    
    new create(env : Env) =>
        _env = env
        let x : U64 = @sfVector2f_to_U64(vec)
        @PrintStruct(x)
```
This will print `1.000000, 2.000000`. Another way to accomplish this hack would be to something like
```pony
type SFVector2fRaw is NullablePointer[SFVector2f]
		...
	    var tmp : U64 = 0
	    @memcpy(addressof tmp, SFVector2fRaw(vec), USize(8))
	    @PrintStruct(tmp)
```

**Comment by jemc** [comment]
Just to link back to the discussion around this that happened in chat, here's the relevant Zulip thread: https://ponylang.zulipchat.com/#narrow/stream/189985-beginner-help/topic/FFI.20receiving.20a.20struct.20(value).20crashes.2E

A summary of my main thoughts on this issue, as expressed there:

- The best, most robust solution would be to integrate the compiler with `libclang` and give it access to the C headers so that it can reproduce the exactly correct C ABI (leveraging clang's extensive knowledge of various platforms and edge cases). This would be a nontrivial effort, but would allow a lot of other useful FFI features and is considered to be generally desirable if we can make it happen cleanly.

- In my opinion, the best short-term solution to unblock the OP would be to partially revert the change that disallows this - disallow it by default for safety, but allow users like the OP to explicitly enable it as a compilation flag, recognizing that it may not be safe/correct for some C calls or on some platforms, allowing the user to take responsibility for the potential correctness issues that result.

- I don't think that the workarounds allowing the language user to handle byte packing/unpacking themselves are a satisfactory kind of solution - it has the same problems that the ponyc compiler has in not knowing whether in the correct call convention the argument is truly represented as a struct or not. In cases where it is not, there would be the same issues the compiler had. So in these workarounds we'd be asking the user to jump through a bunch of additional hoops for no benefit - the same problems would be there as if we had merely done the processing in ponyc as before.

**Comment by SeanTAllen** [comment]
During sync. Another point that was discussed as an option was that we could do code generation of the boilerplate associated with functionality as it stands.

**Comment by SeanTAllen** [comment]
Note that @Praetonus mentions https://github.com/ponylang/ponyc/pull/2012#issuecomment-313830595 that an interface for clang needed to be exposed and we were doing this because it wasn't exposed. It might be possible that it has since been exposed. 

The last comment is "I'm going to submit a patch to Clang to expose that interface.". I have no idea what happened with that.

**Comment by SeanTAllen** [comment]
We would love it if someone would look into interfacing with clang and report back on the feasibility.

**Comment by stefandd** [comment]
@SeanTAllen: any thoughts coming out of the sync re: the short-term solution proposed by @jemc since that would not require the things to *maybe have happened* over at Clang?

**Comment by SeanTAllen** [comment]
I'm personally more in favor of code generation rather than putting something that is broken back in. However, the something that might have happened, if it happened could be the best (and a fairly quick) path forward. 

More information is needed and whoever undertakes writing an RFC would ideally look into and weigh the pros and cons of each approach and present them in the RFC.

**Comment by SeanTAllen** [comment]
If anyone is interested in the nuance of the conversation so far that happened at the sync, you can download the recording at https://sync-recordings.ponylang.io/r/2020_09_08.m4a.

Getting that nuance would be a good idea for anyone who is interested in writing an RFC for this.

**Comment by SeanTAllen** [comment]
We discussed this briefly at sync and believe that between last week's recording (for nuance) and what is on this issue should be enough for someone to draft an RFC for this. If you are interested in exploring the issues here in, go for it.

**Comment by SeanTAllen** [comment]
We discussed this again today at sync and @ergl is looking into this more and will be working on an RFC.

**Comment by redvers** [comment]
This is probably no longer needed, now that we have the ability to do this in cshims.

**Comment by SeanTAllen** [comment]
Those. Can be reopened if needed.

**Comment by stefandd** [comment]
> This is probably no longer needed, now that we have the ability to do this in cshims.

@redvers I looked at cshims documentation and examples, but couldn't find anything that would suggest this works with cshims: can cshims be used to call a C function that works on a pony-side struct?

**Comment by redvers** [comment]
What you can do is write a cshim that takes a pony struct (ie, a pointer) - and then calls the C-side with it as a bare struct… and vice-versa.

**Comment by SeanTAllen** [comment]
[struct-by-value.zip](https://github.com/user-attachments/files/32143946/struct-by-value.zip)

I asked claude to whip up an example. See attached.

**Comment by stefandd** [comment]
[@SeanTAllen](https://github.com/SeanTAllen) thanks for the code; [@redvers](https://github.com/redvers) thanks for your reply.

The example shows why this RFC should not be closed, not why it should.

1) In the example, Pony calls in with a pointer to the Pony struct, and the C-side needs an extra function to convert it into a value struct for the call that actually needs one — adding allocation and overhead on every call. That's still just a workaround for the missing capability.

2) cShims doesn't address the problem. Hand-writing one per struct-valued C signature was always possible before cShims existed. For a library with many functions it requires writing a wrapper you write and maintain per function -- the antithesis of an FFI. More importantly, [@jemc](https://github.com/jemc) points out that this isn't a safe substitute either: the ABI risk that got tuple-passing removed in the first place. Neither Pony nor a user manually writing a C shim knows whether the platform ABI treats a struct as in-register struct or not!

The reasoning behind this RFC was that a specific FFI capability did exist — passing tuples or arrays of bytes directly into C. It was removed with libclang integration proposed as the eventual fix. The removal was completely reasonable at the time, but the fix was never built, and the capability is still missing. Other compilers with the same problem, e.g. Go, solved it with ABI-agnostic code generation instead of asking users to hand-roll it. [@jemc](https://github.com/jemc) also proposed an interim option (opt-in unsafe flag to re-enable value-struct passing) earlier in the thread — worth revisiting instead of closing this.

**Comment by SeanTAllen** [comment]
Well I just had a lot of fun with clang APIs. Here ya go @stefandd, want to test it out?

https://github.com/ponylang/ponyc/pull/6062
